src/Middleware/RateLimiting/src/RateLimitingMiddleware.cs

```diff
@@ -0,0 +1,148 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Threading.RateLimiting;
+using Microsoft.AspNetCore.Http;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+
+namespace Microsoft.AspNetCore.RateLimiting;
+
+/// <summary>
+/// Limits the rate of requests allowed in the application, based on limits set by a user-provided <see cref="PartitionedRateLimiter{TResource}"/>.
+/// </summary>
+public partial class RateLimitingMiddleware
+{
+
+    private readonly RequestDelegate _next;
+    private readonly RequestDelegate _onRejected;
+    private readonly ILogger _logger;
+    private readonly PartitionedRateLimiter<HttpContext> _limiter;
+    private RateLimitLease? _lease;
+
+    /// <summary>
+    /// Creates a new <see cref="RateLimitingMiddleware"/>.
+    /// </summary>
+    /// <param name="next">The <see cref="RequestDelegate"/> representing the next middleware in the pipeline.</param>
+    /// <param name="loggerFactory">The <see cref="ILoggerFactory"/> used for logging.</param>
+    /// <param name="options">The options for the middleware.</param>
+    public RateLimitingMiddleware(RequestDelegate next, ILoggerFactory loggerFactory, IOptions<RateLimitingOptions> options)
+    {
+        _next = next ?? throw new ArgumentNullException(nameof(next));
+
+        if (options.Value.Limiter == null)
+        {
+            throw new ArgumentException("The value of 'options.Limiter' must not be null.", nameof(options));
+        }
+
+        if (options.Value.OnRejected == null)
+        {
+            throw new ArgumentException("The value of 'options.OnRejected' must not be null.", nameof(options));
+        }
+
+        if (loggerFactory == null)
+        {
+            throw new ArgumentNullException(nameof(loggerFactory));
+        }
+
+        _logger = loggerFactory.CreateLogger<RateLimitingMiddleware>();
+        _limiter = options.Value.Limiter;
+        _onRejected = options.Value.OnRejected;
+    }
+
+    // TODO - EventSource?
```

Comment:
Could be valuable to add EventSource logging to this, part of the next pass?

Reply:
I'm fine with waiting for the next pass. It might be better to add EventSource logging to the rate limiter implemenations, but that might be too low level to be useful if there are many.

Reply:
Is there an issue tracking this yet?

Reply:
https://github.com/dotnet/aspnetcore/issues/41266